### ISSUE TITLE
Add number input component

### ISSUE THREAD

**Issue by frankiereyes** [thesmythgroup/vue-toolkit#18]
## Describe the Feature

Add an input component that restricts the input to numbers. Should also support Firefox.

**Comment by justinlettau** [comment]
Would the input `type="number"` ([docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/number)) not work for this? 

```html
<v-input type="number />
```

If needed, a validator could be used to ensure the end value is only numeric characters (if concerned about certain edge cases). 

https://caniuse.com/input-number

**Comment by frankiereyes** [comment]
We had come across an issue in another project where Firefox allows a user to type/paste non-numeric values into an `input[type=number]` field. The [docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/number#using_number_inputs) page references this issue: [1398528](https://bugzilla.mozilla.org/show_bug.cgi?id=1398528)

The good news is we have a workaround we can port into this project!

**Comment by justinlettau** [comment]
Yeah, [can I use](https://caniuse.com/input-number) notes that issue in the "Known Issues" of browsers, but it indicates it's just on older versions 😕 I assume the your workaround is some kind of keypress logic?  I'm wondering if a validator would solve this issue instead?

**Comment by frankiereyes** [comment]
Unfortunately, this is still happening as of Firefox 92.0.1. Yeah, the workaround uses `KeyboardEvent` and `ClipboardEvent` to prevent input of non-numeric characters. A validator could work if we are willing to allow them to input non-numeric characters.

Tagging @pbredenberg since he requested that I add this issue

**Comment by pbredenberg** [comment]
> Would the input `type="number"` ([docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/number)) not work for this?
> 
> ```
> <v-input type="number />
> ```
> 
> If needed, a validator could be used to ensure the end value is only numeric characters (if concerned about certain edge cases).
> 
> https://caniuse.com/input-number

According to the current state of `v-input`, using the default configuration for HTML input tags will not work, because our component doesn't support it ([it's hardcoded](https://github.com/thesmythgroup/vue-toolkit/blob/main/packages/forms/src/components/input/Input.vue#L3) as `text`), therefore that's why I asked Frankie to add this issue.

**Comment by justinlettau** [comment]
@pbredenberg The default behavior of Vue is to merge all non-prop attributes from the component instance onto the root element of the component template.

https://vuejs.org/v2/guide/components-props.html#Non-Prop-Attributes

So in this case, if you provide a `type` attribute on the component (`<v-input type="number" />`), it will replace the default `type="text"` at runtime (rending `<input type="number />`). An example of this being done is in the AWs Auth components ([example here](https://github.com/thesmythgroup/vue-toolkit/blob/main/packages/aws-auth/src/components/sign-in/SignIn.vue#L18)).

**Comment by pbredenberg** [comment]
> @pbredenberg The default behavior of Vue is to merge all non-prop attributes from the component instance onto the root element of the component template.
> 
> https://vuejs.org/v2/guide/components-props.html#Non-Prop-Attributes
> 
> So in this case, if you provide a `type` attribute on the component (`<v-input type="number" />`), it will replace the default `type="text"` at runtime (rending `<input type="number />`). An example of this being done is in the AWS Auth components ([example here](https://github.com/thesmythgroup/vue-toolkit/blob/main/packages/aws-auth/src/components/sign-in/SignIn.vue#L18)).

Mind. Blown! I have been using Vue for 5 years and I never knew about that!

To be fair though, this behavior is not obvious nor ubiquitous for front-end apps. Also the referenced documentation says this:

> While explicitly defined props are preferred for passing information to a child component, authors of component libraries can’t always foresee the contexts in which their components might be used. That’s why components can accept arbitrary attributes, which are added to the component’s root element.

In other words, the _preferred way_ of handling props is to _explicitly define them_ when it is fathomable that someone might use it in a particular way (in our case, defining the `type` attribute on an element is so common we might consider defining a prop for it).

To me we should help clarify this point by adding examples [to our own documentation](https://thesmythgroup.github.io/vue-toolkit/#/forms/input), and mention that this is how we intend the library to be used, since others who might be unaware of that behavior could open similar issues.

Maybe this could be a good first PR for you @frankiereyes?

**Comment by pbredenberg** [comment]
@frankiereyes please add the documentation that @justinlettau mentions to the form demo in our docs site. Thank you both!

**Comment by frankiereyes** [comment]
@justinlettau Sorry for the delay, I'm getting back to this and noticed that you had made this [commit](https://github.com/thesmythgroup/vue-toolkit/commit/d76611e5a473fe3317dc20d826f03e4a212935a3#diff-46b42b4229cd7a39c564e780bb665a8bde4fdf722007e8473f167fe53ed4b995). Was this intended to show how we expect to override the input type?

**Comment by justinlettau** [comment]
> @justinlettau Sorry for the delay, I'm getting back to this and noticed that you had made this [commit](https://github.com/thesmythgroup/vue-toolkit/commit/d76611e5a473fe3317dc20d826f03e4a212935a3#diff-46b42b4229cd7a39c564e780bb665a8bde4fdf722007e8473f167fe53ed4b995). Was this intended to show how we expect to override the input type?

No, that was a different use case. You can use the `type` attribute on `v-input` the same way you would a native HTML input element.

```html
<v-input type="text" />
<v-input type="email" />
<v-input type="number" />
<v-input type="password" />
<v-input type="tel" />
// etc.
```

**Comment by frankiereyes** [comment]
Ok, great. That's what I understood from the previous conversations. I started updating the Form example and noticed the 'Level' field and thought you might have come up with a different idea.

I'll get back on this tomorrow morning and update the docs. Thanks!!

**Comment by justinlettau** [comment]
Closed via #19
